### PR TITLE
Improving autocompletion for users and channels

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Improving autocompletion for users and channels. Mainly handling the blur event
in the users input field.

#### Ticket Link
[MM-17330](https://mattermost.atlassian.net/browse/MM-17330)
Implementation of the [MM-17505](https://mattermost.atlassian.net/browse/MM-17505) specification (See the linked invisions in the comments).

### CODE REVIEW

**Comment by srkgupta** [comment]
The changes will be tested as part of 5.16 release Guest Account feature. Removing the QA Review label.

**Comment by lieut-data** [comment]
@jespino, I'm having trouble understanding exactly what this PR is about. The linked JIRA is ambiguous as well. Could you post before and after screenshots to help explain?

**Comment by jespino** [comment]
@lieut-data Yes, I updated the description with the real definition of the new behavior.

**Comment by lieut-data** [comment]
Sorry to be a nuisance @jespino, but I'm finding myself really digging through the linked JIRA issues and the Invision diagrams to understand what this change is about. I think it would be really useful if the description of all PRs is self-contained. 

If I understand correctly, the commend on this PR could be:
> The guest accounts feature includes an autocomplete input field to simplify finding a user to invite. Sometimes there is no match, but the implementation was incorrectly discarding that part of the input when the element lost focus. Fix this to preserve the input, so we can surface the potential mistake on the next screen of the modal.

Is this accurate? Can you update the description accordingly?

File: components/widgets/inputs/users_emails_input.jsx [hunk 1]
```diff
@@ -197,13 +258,14 @@ export default class UsersEmailsInput extends React.Component {
...
                 styles={this.customStyles}
                 onChange={this.onChange}
-                loadOptions={this.props.usersLoader}
-                isValidNewOption={isEmail}
+                loadOptions={this.optionsLoader}
+                isValidNewOption={this.showAddEmail}
                 isMulti={true}
                 isClearable={false}
-                className='UsersEmailsInput'
+                className={'UsersEmailsInput ' + (this.props.inputValue === '' ? 'empty' : '')}
```
Comment (lieut-data) [inline components/widgets/inputs/users_emails_input.jsx]: Could we use the `classnames` import to simplify this logic, and make the resulting classname more explicit (and greppable!)

File: components/widgets/inputs/channels_input.jsx [hunk 2]
```diff
@@ -109,15 +146,20 @@ export default class ChannelsInput extends React.Component {
...
             <AsyncSelect
+                ref={this.selectRef}
                 styles={this.customStyles}
                 onChange={this.onChange}
-                loadOptions={this.props.channelsLoader}
+                loadOptions={this.optionsLoader}
                 isMulti={true}
                 isClearable={false}
-                className='ChannelsInput'
+                className={'ChannelsInput ' + (this.props.inputValue === '' ? 'empty' : '')}
```
Comment (lieut-data) [inline components/widgets/inputs/channels_input.jsx]: See later comment re: `classnames`.

File: components/invitation_modal/invitation_modal_guests_step.jsx [hunk 3]
```diff
@@ -92,7 +100,7 @@ export default class InvitationModalGuestsStep extends React.Component {
             return this.props.myInvitableChannels;
         }
         return this.props.myInvitableChannels.filter((channel) => {
-            return channel.display_name.toLowerCase().indexOf(value.toLowerCase()) === 0;
+            return channel.display_name.toLowerCase().indexOf(value.toLowerCase()) === 0 || channel.name.toLowerCase().indexOf(value.toLowerCase()) === 0;
```
Comment (lieut-data) [inline components/invitation_modal/invitation_modal_guests_step.jsx]: nit: use `startsWith` for clarity?

File: components/invitation_modal/invitation_modal_guests_step.jsx [hunk 4]
```diff
@@ -40,22 +40,34 @@ export default class InvitationModalGuestsStep extends React.Component {
...
             channels: props.defaultChannels || [],
+            usersInputValue: '',
+            channelsInputValue: '',
         };
     }
 
     onUsersEmailsChange = (usersAndEmails) => {
         this.setState({usersAndEmails});
-        this.props.onEdit(usersAndEmails.length > 0 || this.state.channels.length > 0 || this.state.customMessage !== '');
+        this.props.onEdit(usersAndEmails.length > 0 || this.state.channels.length > 0 || this.state.customMessage !== '' || this.state.usersInputValue || this.state.channelsInputValue);
```
Comment (lieut-data) [inline components/invitation_modal/invitation_modal_guests_step.jsx]: Using the force here, but there's a ton of internal state for this component. Entirely out of scope for the PR, but I think we'd be better served by using Redux for complex components like this, improving testability.

File: components/widgets/inputs/users_emails_input.jsx (hunk 1 above)
Comment (jespino) [reply components/widgets/inputs/users_emails_input.jsx]: I didn't know classnames, but yes, sure we can :)

File: components/invitation_modal/invitation_modal_guests_step.jsx (hunk 4 above)
Comment (jespino) [reply components/invitation_modal/invitation_modal_guests_step.jsx]: Yes, this have a lot of internal state related to the input fields in the form, we can move that to redux, but is somehow ephemeral data that, from my point of view, makes no sense to expose into the redux store. Is just local state.

File: components/widgets/inputs/users_emails_input.jsx (hunk 1 above)
Comment (jespino) [reply components/widgets/inputs/users_emails_input.jsx]: Done

File: components/invitation_modal/invitation_modal_guests_step.jsx (hunk 3 above)
Comment (jespino) [reply components/invitation_modal/invitation_modal_guests_step.jsx]: Done

File: components/widgets/inputs/channels_input.jsx (hunk 2 above)
Comment (jespino) [reply components/widgets/inputs/channels_input.jsx]: Done

File: components/widgets/inputs/users_emails_input.jsx [hunk 5]
```diff
@@ -188,10 +199,57 @@ export default class UsersEmailsInput extends React.Component {
...
+            });
+
+            for (const option of this.state.options) {
+                if (this.props.inputValue === option.username || this.props.inputValue === ('@' + option.username)) {
+                    this.onChange([...values, option]);
+                    this.props.onInputChange('');
+                    return;
+                } else if (this.props.inputValue === option.email) {
+                    this.onChange([...values, option]);
+                    this.props.onInputChange('');
```
Comment (marianunez) [inline components/widgets/inputs/users_emails_input.jsx]: Why two conditions it seems that they both execute the same code block in each case? 🤔

**Pull request merged**